File: 347. Top K Frequent Elements/Main.java
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;
import java.util.Map.Entry;

class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] result = solution.topKFrequent(new int[]{1,2}, 2);
        for (int el : result) {
            System.out.println(el);
        }
    }
}

class Solution {
    /**
     * Measure count of each element.
     * Fill a bucket of list based on the count of each element.
     * Go over the bucket in descending order till you have k elements poped.
     */
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> countMap = new HashMap<>();
        for (int el : nums) {
            countMap.put(el, countMap.getOrDefault(el, 0) + 1);
        }

        Stack<Integer>[] buckets = new Stack[nums.length + 1];
        for (Entry<Integer, Integer> entry : countMap.entrySet()) {
            Stack<Integer> bucket = buckets[entry.getValue()];
            if (bucket == null) {
                bucket = new Stack<>();
                buckets[entry.getValue()] = bucket;
            }
            bucket.push(entry.getKey());
        }

        int[] result = new int[k];
        int j = 0;
        for (int i = nums.length; i >= 0; i--) {
            Stack<Integer> bucket = buckets[i];
            while (bucket != null && !bucket.isEmpty()){
                result[j] = bucket.pop();
                j++;
                if (j == k) return result;
            }
        }

        return result;
    }
}

File: 347. Top K Frequent Elements/main.go
package main

import "fmt"

func main() {
	fmt.Println(topKFrequent([]int{1, 2}, 2))
}

/*
Measure count of each element.
Fill a bucket of list based on the count of each element.
Go over the bucket in descending order till you have k elements poped.
*/
func topKFrequent(nums []int, k int) []int {
	countMap := make(map[int]int)

	for _, el := range nums {
		countMap[el]++
	}

	buckets := make([][]int, len(nums)+1)
	for key, value := range countMap {
		buckets[value] = append(buckets[value], key)
	}

	var topK []int
	for i := len(buckets) - 1; i >= 0; i-- {
		for j := 0; j < len(buckets[i]); j++ {
			if k <= 0 {
				return topK
			}
			topK = append(topK, buckets[i][j])
			k--
		}
	}

	return topK
}
